### DISCUSSION TITLE
Persist layout during navigation so that it is not re-rendered

### DISCUSSION THREAD

**Discussion by kcoet** [vikejs/vike#1284] [Discussion]
Is there a way to keep the layout from re-rendering when navigating between pages? I have tried several ways recommended in the documentation, none of them worked. There was a section in my layout that needed to fetch data, now the data is refetched every time it navigates.

Here's a simple example: https://stackblitz.com/edit/stackblitz-starters-uet6jt?file=renderer%2FPageShell.tsx

**Comment by brillout** [comment]
You're using React so re-rendering the whole Layout/Page is actually idiomatic React usage. (That's why React has a virtual dom.)

As for nested views (aka Nested Layouts), it's work-in-progress, see https://github.com/vikejs/vike/issues/346. The data of nested views will only be fetched if the view changes. (Btw. I'd be up for a sponsoring deal to increase the priority of this ticket, in case that's an argument you can communicate to management.)

In the meantime, would using a cache be a satisfying workaround for you?

**Reply by kcoet** [reply]
> You're using React so re-rendering the whole Layout/Page is actually idiomatic React usage. (That's why React has a virtual dom.)

I used the Apollo template example, this resulted in flickering, data refetching and loading returning `true`. I think, in Remix I don't experience this.

**Reply by kcoet** [reply]
If I inject the Apollo Initial State directly into the HTML without going through the `pageContext` (as below), it works fine, there is no flickering and the data is not refetched.

```js
const documentHtml = escapeInject`<!DOCTYPE html>
    <html>
      <body>
        <div id="page-content">${dangerouslySkipEscape(pageHtml)}</div>
        <script>
        window.__APOLLO_STATE__=${dangerouslySkipEscape(JSON.stringify(apolloIntialState).replace(/</g, '\\u003c'))};
        </script>
      </body>
    </html>
```

**Reply by brillout** [reply]
AFAICT you can achieve the same using `pageContext`.

Btw. we're working on [new integration pacakges](https://github.com/vikejs/vike/issues/1263) &mdash; a `vike-react-apollo` would be really neat. Contribution welcome in case that's something you'd be up to.
